ARCs/arc-0004.md

```diff
@@ -0,0 +1,498 @@
+* If `x` is a boolean value `bool`:
+    * `enc(x)` is a single byte whose **most significant bit** is either 1 or 0, if `x` is true or false respectively. All other bits are 0. Note: this means that a value of true will be encoded as `0x80` (`10000000` in binary) and a value of false will be encoded as `0x00`. This is in contrast to most other encoding schemes, where a value of true is encoded as `0x01`.
+
+Since `string` and `address` are aliases for `byte[]` and `byte[32]`
+respectively (and `byte` is an alias for `uint8`), the rules for
+encoding these types are covered above.
+
+## Further considerations
+
+This document describes argument and return value encoding only.
+There are several other aspects to invoking an Algorand app properly
+that require standardization to improve high-level access to
+contracts.
+
+### Opt In, Close Out, and Clear State.
+
+An ARC-4 app **MUST** allow any account to opt-in and close-out, using
+an application call transaction with no arguments an OnComplete equal
+to . Clients **MAY** determine that such an opt-in is unnecessary by
+querying the blockchain for the app's schema. Opt-in is unnessary if
+the app has no local schema.  However, even if a local schema is
+present, some methods may be callable without opt-in. Clients **MAY**
+attempt method calls without opt-in.
+
+Clients **MAY** display the minimum balance increment that such an
+opt-in would require.  If an opt-in or close-out contains arguments,
+an ARC-4 app **MUST** fail the transaction, or interpret the arguments
+as an invocation as described in the rest of this document, to perform
+along with the opt-in or close-out.
+
+### Foreign apps, asas, and accounts
+
+Some AVM opcodes require specific values to be placed in the "foreign
+arrays" of the application call transaction. Three special types,
+`account`, `asset`, `application` allow methods to describe these
+requirements. To encode method calls with such types, place the value
+in question in the Accounts (`apat`), Foreign Assets (`apas`), or
+Foreign Assets (`apfa`) arrays, and encode a uint8 referring to that
+value in the normal location for the value. For the Accounts and
+Applications array, a 0 implicitly refers to the Sender and the called
+application respectively, and the passed values begin at 1.
+
+When passing addresses, ASAs, or apps that are _not_ required to be
+accessed by such opcodes, ARC-4 Apps **SHOULD** use the base types for
+passing these types: `address` and `uint64`.
+
+### Transaction groups.
+
+Some apps require that they are invoked as part of a larger
+transaction group, containing specific additional transactions. Future
+versions of this document may describe how a method signature could
+describe these requirements.
+
+ARC-4 Apps **SHOULD** be constructed to allow their invocations to be
+combined in a single atomic group if they can do so safely. For
+example, they should use `txns` to examine the previous index for a
```

Comment:
The word "previous" is confusing here, because it is the first time that the TEAL stack is acknowledged.

Reply:
I'm actually referring to the previous transaction index here, in the group.  Your point that it needs clarity is taken.